### ISSUE TITLE
Failed to build flashinfer from source

### ISSUE THREAD

**Issue by peng1999** [flashinfer-ai/flashinfer#432]
Steps to reproduce:

```sh
git clone https://github.com/flashinfer-ai/flashinfer.git --recursive
cd flashinfer
python -m venv venv
. venv/bin/activate
python -m pip install -v -e python
```

Output:

```
Using pip 24.2 from /mnt/workspace/pgw/flashinfer/venv/lib/python3.10/site-packages/pip (python 3.10)
Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple
Obtaining file:///mnt/workspace/pgw/flashinfer/python
  Running command pip subprocess to install build dependencies
  Using pip 24.2 from /mnt/workspace/pgw/flashinfer/venv/lib/python3.10/site-packages/pip (python 3.10)
  Looking in indexes: https://pypi.tuna.tsinghua.edu.cn/simple
  Collecting setuptools>=40.8.0
    Using cached https://pypi.tuna.tsinghua.edu.cn/packages/e1/58/e0ef3b9974a04ce9cde2a7a33881ddcb2d68450803745804545cdd8d258f/setuptools-72.1.0-py3-none-any.whl (2.3 MB)
  Installing collected packages: setuptools
  Successfully installed setuptools-72.1.0
  Installing build dependencies ... done
  Running command Checking if build backend supports build_editable
  Checking if build backend supports build_editable ... done
  Running command Getting requirements to build editable
  Traceback (most recent call last):
    File "/mnt/workspace/pgw/flashinfer/venv/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
      main()
    File "/mnt/workspace/pgw/flashinfer/venv/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/mnt/workspace/pgw/flashinfer/venv/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 132, in get_requires_for_build_editable
      return hook(config_settings)
    File "/tmp/pip-build-env-wt9kme3x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 458, in get_requires_for_build_editable
      return self.get_requires_for_build_wheel(config_settings)
    File "/tmp/pip-build-env-wt9kme3x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 327, in get_requires_for_build_wheel
      return self._get_build_requires(config_settings, requirements=[])
    File "/tmp/pip-build-env-wt9kme3x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 297, in _get_build_requires
      self.run_setup()
    File "/tmp/pip-build-env-wt9kme3x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 497, in run_setup
      super().run_setup(setup_script=setup_script)
    File "/tmp/pip-build-env-wt9kme3x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 313, in run_setup
      exec(code, locals())
    File "<string>", line 28, in <module>
  ModuleNotFoundError: No module named 'torch'
  error: subprocess-exited-with-error

  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> See above for output.

  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /mnt/workspace/pgw/flashinfer/venv/bin/python /mnt/workspace/pgw/flashinfer/venv/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py get_requires_for_build_editable /tmp/tmpmj9ddw26
  cwd: /mnt/workspace/pgw/flashinfer/python
  Getting requirements to build editable ... error
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip
```

**Comment by yzh119** [comment]
you have to build flashinfer under a python environment where pytorch is installed.
See: https://docs.flashinfer.ai/installation.html#compile-from-source

**Comment by peng1999** [comment]
The error still exists after pytorch being installed.

```console
# python -c "import torch; print(torch.__version__, torch.version.cuda)"
2.4.0+cu121 12.1
# python -m pip install -v -e python
... (the same error as before)
```

I noticed that the error happened in a new envirment located in `/tmp/pip-build-env-wt9kme3x/overlay/`. Maybe it is related to some new behaviour in recent versions of pip?

**Comment by yzh119** [comment]
It's looks weird to me.
FYI, we fixed the setuptools version to 69.5.1 in our CI: https://github.com/flashinfer-ai/flashinfer/blob/ea0ba9a51238597bd7863b6e3c9bfda574df4df5/scripts/run-ci-build-wheel.sh#L39

The v0.1.4 release is running: https://github.com/flashinfer-ai/flashinfer/actions/runs/10316658694, new wheels should be available several hours later.

**Comment by peng1999** [comment]
Thanks for pointing out the CI script. I think the key point is the `--no-isolation` flag here:

https://github.com/flashinfer-ai/flashinfer/blob/ea0ba9a51238597bd7863b6e3c9bfda574df4df5/scripts/run-ci-build-wheel.sh#L47

Using this command, I can start building the repo.

The `pip install` flag is `--no-build-isolation`.

Closing the issue, but it's probably worth mentioning this flag in the README.

cc @Brennanzuz from #316
